src/wasm.h

```diff
@@ -88,25 +88,37 @@ struct Name : public cashew::IString {
   }
 };
 
-// An index in a wasm module
-typedef uint32_t Index;
-
-// An address in linear memory. For now only wasm32
-struct Address {
-  typedef uint32_t address_t;
-  address_t addr;
-  Address() : addr(0) {}
-  Address(uint64_t a) : addr(static_cast<address_t>(a)) {
-    assert(a <= std::numeric_limits<address_t>::max());
-  }
-  Address& operator=(uint64_t a) {
-    assert(a <= std::numeric_limits<address_t>::max());
-    addr = static_cast<address_t>(a);
+// An index in a wasm module. This could be e.g. a local index or an index in
+// linear memory (i.e. an address).
+template <typename T>
+struct WasmIndex {
+  typedef T value_type;
+  T addr;
+  WasmIndex() : addr(0) {}
+  static void CheckValid(uint64_t v) {
+    // Allow -1 as an invalid sentinel value
+    assert(v <= std::numeric_limits<T>::max() ||
+	   v == static_cast<uint64_t>(-1));
+  }
+  WasmIndex(uint64_t a) : addr(static_cast<T>(a)) {
```

Comment:
Perhaps we should add constructors to `Literal` for `size_t`, if that works? And do the same here? Overall `Literal` and this class have the same problem, so it would be consistent to solve it in the same way.

I think literal 0 should be ok, if we do what `Literal` does, which is have a constructor for all `[u]int[32|64]_t`. `int` has to be done of those. I'm not sure if `size_t` is, I guess we can just try it and see (but given I don't think we have warnings on `Literal` usage, I tend to think it's already doing the right thing).